File: backend/src/schemas/manual.py
"""Pydantic schemas for manual debug mode."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


# ============================================================================
# Sessions
# ============================================================================


class ManualSessionCreate(BaseModel):
    """Request to create a manual session."""
    profile_id: UUID
    name: str = Field(max_length=255)
    description: str | None = None


class ManualSessionUpdate(BaseModel):
    """Request to update a manual session."""
    name: str | None = Field(None, max_length=255)
    description: str | None = None
    status: str | None = Field(None, pattern=r"^(active|archived)$")
    profile_snapshot: dict[str, Any] | None = None


class ManualSessionResponse(BaseModel):
    """Response for a manual session."""
    id: UUID
    profile_id: UUID
    profile_snapshot: dict[str, Any]
    name: str
    description: str | None
    status: str
    created_at: datetime
    updated_at: datetime


class ManualSessionListResponse(BaseModel):
    """Response for session listing."""
    sessions: list[ManualSessionResponse]
    total: int


# ============================================================================
# Prompt Versions
# ============================================================================


class PromptVersionCreate(BaseModel):
    """Request to create a new prompt version."""
    prompt_text: str
    change_description: str | None = None


class PromptVersionResponse(BaseModel):
    """Response for a prompt version."""
    id: UUID
    step_name: str
    version: int
    prompt_text: str
    change_description: str | None
    is_baseline: bool
    created_at: datetime


class PromptStepSummary(BaseModel):
    """Summary of prompts for a step — latest version info."""
    step_name: str
    latest_version: int
    latest_prompt_id: UUID
    is_baseline: bool
    created_at: datetime


class PromptListResponse(BaseModel):
    """Response for prompt listing (all steps with latest version)."""
    steps: list[PromptStepSummary]


# ============================================================================
# Step Runs
# ============================================================================


class StepRunRequest(BaseModel):
    """Request to run a step."""
    prompt_version_id: UUID | None = None
    custom_prompt: str | None = None
    input_data: dict[str, Any] | None = None
    llm_params: dict[str, Any] | None = None
    run_preprocessors: bool = True
    run_postprocessors: bool = True
    use_mock: bool = True


class StepRunResponse(BaseModel):
    """Response for a step run."""
    id: UUID
    session_id: UUID
    step_name: str
    run_number: int
    prompt_version_id: UUID | None
    rendered_prompt: str | None
    input_data: dict[str, Any] | None
    profile_variables: dict[str, Any] | None
    llm_params: dict[str, Any] | None
    raw_response: str | None
    parsed_result: dict[str, Any] | None
    parse_error: str | None
    tokens_used: int | None
    duration_ms: float | None
    status: str
    preprocessor_results: list[dict[str, Any]] | None
    postprocessor_results: list[dict[str, Any]] | None
    auto_evaluation: dict[str, Any] | None
    llm_judge_evaluation: dict[str, Any] | None
    user_rating: int | None
    user_notes: str | None
    created_at: datetime


class StepRunSummary(BaseModel):
    """Summary of a step run for listing."""
    id: UUID
    run_number: int
    status: str
    duration_ms: float | None
    tokens_used: int | None
    user_rating: int | None
    created_at: datetime


class StepStatusResponse(BaseModel):
    """Status of all steps in a session."""
    steps: dict[str, dict[str, Any]]
    # e.g. {"B1_validate": {"last_run": {...}, "run_count": 3, "status": "completed"}}


class UserRatingUpdate(BaseModel):
    """Request to rate a step run."""
    user_rating: int | None = Field(None, ge=1, le=5)
    user_notes: str | None = None


# ============================================================================
# Processors
# ============================================================================


class ProcessorInfo(BaseModel):
    """Info about an available processor."""
    name: str
    type: str
    applicable_steps: list[str]
    description: str


class ProcessorConfigItem(BaseModel):
    """Config for one processor on a step."""
    processor_name: str
    processor_type: str
    execution_order: int = 0
    enabled: bool = True
    config_params: dict[str, Any] | None = None


class ProcessorConfigUpdate(BaseModel):
    """Request to set processor configs for a step."""
    processors: list[ProcessorConfigItem]


class ProcessorConfigResponse(BaseModel):
    """Response for processor config."""
    step_name: str
    processors: list[ProcessorConfigItem]

File: frontend/src/types/manual.ts
/**
 * TypeScript types for manual debug mode
 */

export interface ManualSession {
  id: string
  profile_id: string
  profile_snapshot: Record<string, unknown>
  name: string
  description: string | null
  status: 'active' | 'archived'
  created_at: string
  updated_at: string
}

export interface ManualSessionListResponse {
  sessions: ManualSession[]
  total: number
}

export interface PromptVersion {
  id: string
  step_name: string
  version: number
  prompt_text: string
  change_description: string | null
  is_baseline: boolean
  created_at: string
}

export interface PromptStepSummary {
  step_name: string
  latest_version: number
  latest_prompt_id: string
  is_baseline: boolean
  created_at: string
}

export interface StepRunRequest {
  prompt_version_id?: string | null
  custom_prompt?: string | null
  input_data?: Record<string, unknown> | null
  llm_params?: Record<string, unknown> | null
  run_preprocessors?: boolean
  run_postprocessors?: boolean
  use_mock?: boolean
}

export interface StepRunResponse {
  id: string
  session_id: string
  step_name: string
  run_number: number
  prompt_version_id: string | null
  rendered_prompt: string | null
  input_data: Record<string, unknown> | null
  profile_variables: Record<string, unknown> | null
  llm_params: Record<string, unknown> | null
  raw_response: string | null
  parsed_result: Record<string, unknown> | null
  parse_error: string | null
  tokens_used: number | null
  duration_ms: number | null
  status: 'pending' | 'running' | 'completed' | 'failed'
  preprocessor_results: ProcessorRunResult[] | null
  postprocessor_results: ProcessorRunResult[] | null
  auto_evaluation: Record<string, unknown> | null
  llm_judge_evaluation: Record<string, unknown> | null
  user_rating: number | null
  user_notes: string | null
  created_at: string
}

export interface StepRunSummary {
  id: string
  run_number: number
  status: string
  duration_ms: number | null
  tokens_used: number | null
  user_rating: number | null
  created_at: string
}

export interface StepStatus {
  run_count: number
  status: string
  last_run_id: string | null
  last_rating: number | null
}

export interface StepStatusResponse {
  steps: Record<string, StepStatus>
}

export interface ProcessorRunResult {
  name: string
  passed: boolean
  output?: Record<string, unknown>
  message?: string
  error?: string
}

export interface ProcessorInfo {
  name: string
  type: 'pre' | 'post'
  applicable_steps: string[]
  description: string
}

export interface ProcessorConfigItem {
  processor_name: string
  processor_type: string
  execution_order: number
  enabled: boolean
  config_params: Record<string, unknown> | null
}

export const STEP_NAMES = [
  'B1_validate',
  'B2_competencies',
  'B3_ksa_matrix',
  'B4_learning_units',
  'B5_hierarchy',
  'B6_problem_formulations',
  'B7_schedule',
  'B8_validation',
] as const

export const STEP_SHORT_NAMES: Record<string, string> = {
  B1_validate: 'B1',
  B2_competencies: 'B2',
  B3_ksa_matrix: 'B3',
  B4_learning_units: 'B4',
  B5_hierarchy: 'B5',
  B6_problem_formulations: 'B6',
  B7_schedule: 'B7',
  B8_validation: 'B8',
}

export const STEP_DESCRIPTIONS: Record<string, string> = {
  B1_validate: 'Валидация и обогащение профиля',
  B2_competencies: 'Формулировка компетенций',
  B3_ksa_matrix: 'KSA-матрица',
  B4_learning_units: 'Учебные единицы',
  B5_hierarchy: 'Иерархия и уровни',
  B6_problem_formulations: 'Формулировки проблем (PBL)',
  B7_schedule: 'Сборка расписания',
  B8_validation: 'Валидация трека',
}
